nodeJS/express/introduction_to_express.md

```diff
@@ -0,0 +1,98 @@
+
+Going forward, we will be diving into how we can use Express in various ways, such as to create a full stack application utilising the `Model View Controller` (MVC) pattern, as well as to create a REST API just like ones you will have used before for things such as the Weather App project or the React Shopping Cart. Along the way, we will be learning about things like databases, authentication, and much more. There is a lot to take in, so take it steady and do not be afraid to ask our community for help in [our Discord server](https://discord.gg/theodinproject)!
+
+### Lesson overview
+
+This section contains a general overview of topics that you will learn in this lesson.
+
+- Describe what Express is.
+- Describe the relationship between Express and Node.
+- Set up a basic Express server.
+- Describe how an incoming request is processed and ultimately responded to by the server.
+- Describe what a middleware function is.
+- Describe how to use `nodemon` to automatically restart your application when a change is made.
+
+### Setting up Express
+
+Let's make a basic Express app. Inside a new directory, start by running `npm init` and filling in each option to create a `package.json`. Once that's created, we can install the Express dependency.
+
+```bash
+npm install express
+```
+
+We can now create an `app.js` file that will serve as the starting point for our Express server. We could name this file anything we want, such as `index.js` or `main.js`, but we will use the same name used in the Express documentation. Inside `app.js`, we will have the following:
+
+```javascript
+const express = require("express");
+const app = express();
+
+app.get("/", (req, res) => res.send("Hello, world!"));
+
+const PORT = 3000;
+app.listen(PORT, () => console.log(`My first Express app - listening on port ${PORT}!`));
+```
+
+Let's break this down. Firstly, we import `express` using `require()`, then we call it to create the actual server that we store in the `app` variable. While you can technically do this here in one line via `const app = require("express")();`, very often you will need features that come from properties on `express`. Many of these features will be covered in later lessons.
+
+We then have a `route handler` - the line beginning with `app.get`. We will come back to this in a moment.
+
+Finally, we tell our server to listen for requests on port 3000 of our [localhost](https://simple.wikipedia.org/wiki/Localhost) (which is basically just the computer's local connection). 3000 is an arbitrary and somewhat conventional default choice, but you can technically use any unused port (e.g. Vite's dev server uses port 5173 by default). Back in your terminal, if you enter `node app.js` then all being well, you should see `My first Express app - listening on port 3000!` logged. Congratulations! Your first Express server is now running.
```

Comment:
What about adding in a lesson note titled "The Port Variable". It says something like "`port` is usually a environment variable and we'll cover environment variables later. For now, it's perfectly fine to have it as a global variable."

Reply:
A lesson note sounds good. Might help throw in some visual variance as well.

Reply:
I've been thinking about this. I'm inclined towards covering environment variables earlier, since learners will be dealing with them during deployments.

I think we could even cover env vars before the Express section altogether.

There's also a relevant issue on it https://github.com/TheOdinProject/curriculum/issues/26851

Reply:
Environment variables being covered before Express does make a heap load of sense, and I'm very on board with that idea.
One of the problems with the current order is that the Local Library tutorial doesn't expect you to push to a repo until the end, so it instructs you to hardcode a DB connection string. Only when it reaches the end with deployment does it tell you to swap it for an env variable, push to GitHub, then set the env value in the host. But then people will often push to a repo only to discover later they probably weren't supposed to, and now have their connection string publicly visible. So .env mention is being brought forwards as a bandaid, but then that makes the security lesson kinda moot since most of it is about .env ...

For now, how would you prefer me to handle env variables in this lesson? Make the wording change with the assumption that env has already been covered pre-Express? Then it'll be resolved when we do get to introducing it in the pre-Express section in a later milestone.

Reply:
Yeah, we can assume learners know about env vars in this lesson.